aesd-char-driver/main.c

```diff
@@ -207,12 +211,119 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
 	return newline_pos;
 }
 
+long aesd_ioctl(struct file *filp, unsigned int cmd, unsigned long arg){
+
+	// Get device
+	struct aesd_dev *dev = filp->private_data;
+	if (dev == NULL) {
+		PDEBUG("Device cannot be accessed");
+		return -EINVAL;
+	}
+	
+	if (cmd == AESDCHAR_IOCSEEKTO){
+		
+		PDEBUG("IOCTL Detected");
+		
+		// Try to lock the device mutex
+		int res = mutex_lock_interruptible(&dev->mutex);
+		if (res != 0){
+			PDEBUG("Unable to lock device mutex");
+			return -ERESTARTSYS;
+		}		
+		
+		// Define type of seek performed on the aesdchar driver
+		struct aesd_seekto seek_details;
+		
+		//Copy user arg into seekto struct
+		unsigned long bytes_not_copied = copy_from_user(&seek_details, (const void __user *)arg, sizeof(seek_details));
+		
+		if (bytes_not_copied != 0) {
+			PDEBUG("Unable to copy arg from user space");
+			mutex_unlock(&dev->mutex);
+			return -EFAULT;
+		}
+		
+		// If write cmd out of range of the number of write commands, error
+		if (seek_details.write_cmd > AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED) {
+			PDEBUG("Out of range write command");
+			mutex_unlock(&dev->mutex);
+			return -EINVAL; 
+		}
+		// If write cmd offset out of range of the command length, error
+		else if (seek_details.write_cmd_offset > dev->buffer.entry[seek_details.write_cmd].size) {
+			PDEBUG("Out of range write offset");
+			mutex_unlock(&dev->mutex);
+			return -EINVAL;
+		}
+		
+		PDEBUG("seek command: %d, %d", seek_details.write_cmd, seek_details.write_cmd_offset);
+		
+		// Seek to start of command and add the offset within the command
+		loff_t new_offset = 0;
+		for (int i = 0; i < seek_details.write_cmd; i++) {
+			new_offset += dev->buffer.entry[i].size;
+		}
+		new_offset += seek_details.write_cmd_offset;
+		
+		// Update file position pointer to new offset
+		filp->f_pos = new_offset;
+		mutex_unlock(&dev->mutex);
+		return 0;
+	}
+	else {
+		mutex_unlock(&dev->mutex);
+		return -ENOTTY;
+	}
+}
+
+loff_t aesd_llseek(struct file *filp, loff_t offset, int whence){
+
+	PDEBUG("llseek Triggered");
+
+	// Get device
+	struct aesd_dev *dev = filp->private_data;
+	if (dev == NULL) {
+		PDEBUG("Device cannot be accessed");
+		return -EINVAL;
+	}
+	
+	// Try to lock the device mutex
+	int res = mutex_lock_interruptible(&dev->mutex);
+	if (res != 0){
+		PDEBUG("Unable to lock device mutex");
+		return -ERESTARTSYS;
+	}	
+	
+	// Total size of all entries in the buffer
+	loff_t total_size = 0;
+     	struct aesd_buffer_entry *entry;
+	uint8_t index = 0;
+     	AESD_CIRCULAR_BUFFER_FOREACH(entry, &dev->buffer, index){
+     		if(entry->buffptr){
+    			total_size += entry->size;
+		}
+	}
+	
+	PDEBUG("total_size: %d", total_size);
+	
+	// Reposition offset for fixed-sized device 
+	loff_t fixed_size_offset = fixed_size_llseek(filp, offset, whence, total_size);
+	
+	PDEBUG("llseek offset: %d", fixed_size_offset);
+	
+	mutex_unlock(&dev->mutex);
```

Comment:
@spinoandraptos you could move the mutex_unlock before call to fixed_size_llseek here.